src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -435,12 +435,16 @@ bool ObjectMonitor::enter(JavaThread* current) {
 
     // We cleared the pending monitor info since we've just gotten past
     // the enter-check-for-suspend dance and we now own the monitor free
-    // and clear, i.e., it is no longer pending. The ThreadBlockInVM
-    // destructor can go to a safepoint at the end of this block. If we
+    // and clear, i.e., it is no longer pending.
+    // We can go to a safepoint at the end of this block. If we
     // do a thread dump during that safepoint, then this thread will show
     // as having "-locked" the monitor, but the OS and java.lang.Thread
     // states will still report that the thread is blocked trying to
     // acquire it.
+
+    // Completed the tranisition.
+    SafepointMechanism::process_if_requested(current);
+    current->set_thread_state(_thread_in_vm);
```

Comment:
I feel very uncomfortable that we remain _thread_blocked_trans across such a lengthy chunk of code - particularly the call to exit()! This is an abuse of the trans states which are only supposed to exist and be used to ensure the correctness of the Dekker-duality when setting and reading the thread state.

And I still would prefer to see these state changes and related safepoint-mechanism checks encapsulated somehow.

Reply:
Yes we should figure out something here.
Note that we use to call exit() while blocked.